libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,185 @@
+import android.os.Handler;
+import android.os.Looper;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.mapbox.android.core.location.LocationEngine;
+import com.mapbox.android.core.location.LocationEngineCallback;
+import com.mapbox.android.core.location.LocationEngineRequest;
+import com.mapbox.android.core.location.LocationEngineResult;
+import com.mapbox.geojson.Point;
+import com.mapbox.navigator.FixLocation;
+import com.mapbox.navigator.NavigationStatus;
+import com.mapbox.services.android.navigation.v5.internal.navigation.MapboxNavigator;
+
+import java.io.File;
+import java.util.Date;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import timber.log.Timber;
+
+class EnhancedLocationEngineDecorator implements LocationEngine {
+  private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+  private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = 500;
+  private final LocationEngine locationEngine;
+  private final MapboxNavigator mapboxNavigator;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private final ScheduledExecutorService executorService;
+  private Location rawLocation;
+  private Location enhancedLocation;
+
+  EnhancedLocationEngineDecorator(@NonNull LocationEngine locationEngine,
+                                  @NonNull MapboxNavigator mapboxNavigator,
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+    onFreeDrive();
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, null);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
```

Comment:
what happens if this method is called twice?

Reply:
By definition shouldn't be called twice.

Looking at this, I realized that this could be called more than once from `MapboxNavigation`. This needs to be called only once when the service is not bound. Fixing. cc @LukasPaczos 

Do we want to add extra checks to prevent multiple calls to happened?

Reply:
yeah this could result in NPE right?